Psalm integration

konarshankar07 opened the pull request:
Signed-off-by: Shankar Konar <konar.shankar2013@gmail.com>

<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->
Psalm integration added

konarshankar07 committed:
Psalm integration

Signed-off-by: Shankar Konar <konar.shankar2013@gmail.com>

konarshankar07 committed:
Removed redundant psalm config file

Signed-off-by: Shankar Konar <konar.shankar2013@gmail.com>

boesing commented:
Thanks, @konarshankar07 
We have to add the new CI pipeline to this package before we can merge this.
Anyways, thanks for your efforts! This will still count for hacktoberfest due to the `hacktoberfest-accepted` label 👍🏼 

@samsonasik might handle the next steps of this, he might have plans on how to proceed with this component.

konarshankar07 commented:
Hello @boesing 
If you provide the steps to add the new CI pipeline then I can implement that too. I guess you are going to implement the github actions for the CI pipeline
Thanks

froschdesign commented:
@boesing 
The CI workflow is already there, see at the 2.12.x branch:

> This branch is 30 commits ahead, 1 commit behind 3.0.x.

https://github.com/laminas/laminas-view/blob/2.12.x/.github/workflows/continuous-integration.yml

Some cleanup is needed: https://github.com/laminas/laminas-view/issues/71

froschdesign commented:
@konarshankar07 
Please use the 2.12.x as target branch for this pull request. Thanks!

Ocramius commented:
Seems like our own deprecations are at fault with the failure here :|

boesing commented:
> Seems like our own deprecations are at fault with the failure here :|

actually this is due to the phpunit change regarding `deprecationsToExceptions`.
The Tests which fail are to ensure the the deprecation will be triggered (just not that well implemented) and these now fail in latest phpunit.

konarshankar07 commented:
@boesing @Ocramius 
Yes, I have fixed this here https://github.com/laminas/laminas-crypt/pull/14/commits/034976ef0b3097ac88df8487f7cffb0d5a531721 already and will do the same

konarshankar07 committed:
Fixed phpunit test

Signed-off-by: Shankar Konar <konar.shankar2013@gmail.com>

boesing commented on .psalm.xml.dist:
```diff
@@ -0,0 +1,33 @@
+        </ignoreFiles>
+    </projectFiles>
+
+    <issueHandlers>
```
if you add `psalm/plugin-phpunit` and enable it with `vendor/bin/psalm-plugin`, you should be able to remove these issue handlers.

boesing commented on .psalm.xml.dist:
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0"?>
+<psalm
+    totallyTyped="true"
```
`totallyTyped=true` is deprecated in favor of `errorLevel=1`.

konarshankar07 replied:
If I run the pslam with errorLevel=1 then I'm getting 4599 errors. Do we need to fix all the errors? or Am I missing something?
![image](https://user-images.githubusercontent.com/39480008/136919345-03e9fc8e-7e2a-4954-b745-b3101d36f3a2.png)
Thanks

weierophinney replied:
The idea is to establish an initial baseline; see #56 as it details this.

I've done that and pushed changes to the branch at this time.

boesing reviewed (changes requested):
Overall, LGTM - just a few adjustments so we can avoid suppressions.

Another thing is, that it seems that we only have less than 20 errors in this projects. Would you mind fixing these so we end up having 100% typed code rather than suppressing these few errors via baseline? That would really be appreciated 🙏🏼

weierophinney committed:
qa: perform Psalm auto-corrections and establish baseline

This patch runs auto-fixes as suggested by Psalm, and establishes an initial baseline.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: revert BC break in signature

Psalm autofixes added nullable scalar typehint to an argument, and a return type hint, which each would change the signature.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
